src/serac/numerics/functional/tests/functional_nonlinear.cpp

```diff
@@ -69,16 +66,20 @@ void check_gradient(Functional<T>& f, mfem::Vector& U)
   df1 -= f(U_minus);
   df1 /= (2 * epsilon);
 
+  auto [value, dfdU] = f(differentiate_wrt(U));
```

Comment:
I really like this syntax as is also gets rid of the implicit caching and state we had before.

Reply:
there is still some internal state (the derivatives are evaluated at a specific point)-- this change is mostly cosmetic. If the internal state is undesirable to you, we could consider a way to remove it.